File: etl/rueckkehrer.py
"""
Biber und Fischotter — zwei Arten, die zurückgekommen sind.

GEPFLEGTE REIHE. Die Zahlen sind die offiziellen österreichischen
Meldungen nach Artikel 17 der FFH-Richtlinie, abgelesen im
Berichtswerkzeug der Europäischen Umweltagentur. Ein stabiler
Direktlink auf die Zahlen existiert nicht; das Webtool baut die Tabelle
je Land und Periode zur Laufzeit auf.

  Quelle:    Eionet, Article 17 Web Tool — Austria, Mammals
  URL:       https://nature-art17.eionet.europa.eu/article17/species/report/
             ?period=6&group=Mammals&country=AT&region=
             (period 1 = 2001–2006, 3 = 2007–2012, 5 = 2013–2018,
              6 = 2019–2024)
  Gegenprobe Biber:      BOKU, Institut für Wildbiologie und Jagdwirtschaft
  Gegenprobe Fischotter: Land Niederösterreich, Fischotter-Kartierung 2022/23
  Abgerufen: 26.08.2026

WARUM DIESER ABSCHNITT DER GEGENPOL IST

Die übrigen Abschnitte dieses Dashboards zeigen Rückgang. Dieser zeigt,
dass Rückgang kein Naturgesetz ist: Der Biber war in Österreich
ausgerottet — 1869 erlosch der letzte heimische Bestand. Zwischen 1976
und 1982 wurden an drei Stellen Tiere ausgesetzt, an der Donau bei Wien,
in der Ettenau und in den Salzachauen. Der Fischotter war nie ganz weg,
stand aber in den 1970er- und 1980er-Jahren auf seinem Tiefstand und
besiedelte im Jahr 2000 noch rund ein Fünftel der Bundesfläche.

DREI DINGE, DIE BEIM WEITERPFLEGEN AUFFALLEN WERDEN

1. Die Werte sind Spannen, keine Zählungen. Beide Arten werden über
   Reviere und Nachweise erhoben und auf Individuen hochgerechnet. Genau
   deshalb steht im Diagramm die Spanne und nicht ein Mittelwert: eine
   einzelne Zahl wäre eine Genauigkeit, die die Erhebung nicht hergibt.

2. Beim Fischotter fehlt die Periode 2013–2018. Das ist keine Lücke in
   der Natur, sondern ein Einheitenwechsel: Österreich meldete für diese
   Periode keine Individuen, sondern besetzte Rasterzellen zu einem
   Quadratkilometer (alpin 1.555, kontinental 1.354). Diese Zahlen sind
   mit den Individuenzahlen davor und danach nicht vergleichbar und
   werden deshalb NICHT umgerechnet, sondern ausgelassen. Wer sie
   trotzdem einsetzt, erzeugt einen Einbruch, den es nicht gab.

3. Ein Teil des Zuwachses kann bessere Erfassung sein, nicht mehr Tiere.
   Nachgeprüft am 26.08.2026, und der Befund fiel günstiger aus als
   zunächst notiert: In der Meldung 2019–2024 tragen **beide** Arten in
   der **alpinen** Region den Vermerk `genuine` — die Verbesserung des
   Erhaltungszustands von U1 auf FV ist dort als tatsächliche
   Verbesserung gemeldet, nicht als Ergebnis besseren Wissens.
   Kontinental steht `noChange` (Biber) beziehungsweise
   `noChange`/`unknown` (Fischotter).

   Wichtig für die Formulierung: Der Vermerk hängt an der **Bewertung
   des Erhaltungszustands**, nicht an der einzelnen Bestandszahl. Der
   Sprung von rund 8.900 Bibern im Winter 2020/21 (BOKU) auf 13.833 bis
   16.654 bleibt deshalb vorsichtig zu lesen — belegt ist, dass sich der
   Zustand im Alpenraum wirklich verbessert hat, nicht dass jede Zahl
   eins zu eins mehr Tiere bedeutet.

DIE GEGENBEWEGUNG GEHÖRT DAZU

Erholung ist hier nicht das Ende der Geschichte. Seit 2019 gibt es in
mehreren Bundesländern Verordnungen, die eine Entnahme erlauben: beim
Fischotter Niederösterreich höchstens 50 Tiere im Kalenderjahr, die
Steiermark 40, Oberösterreich seit Mitte 2022 im ersten vollen Jahr 64;
beim Biber regelt Niederösterreich seit 2019 Dammentfernung, Lebendfang
und in bestimmten Fällen die Tötung. Das steht in der Hinweiszeile,
damit der Abschnitt nicht als Entwarnung gelesen wird.
"""

from __future__ import annotations

from gemeinsam import log, pflegepruefung, quelle_vermerken, warnen

STAND_JAHR = 2025          # Erscheinungsjahr der Meldung 2019–2024
PERIODE = "2019–2024"

# Das Jahr, in dem der letzte heimische Biberbestand erlosch. Es steht
# nicht auf der Achse — der Abstand zu 2001 wäre auf einer Kategorie-
# achse gleich breit wie sechs Jahre und damit eine Zeitlüge. Es trägt
# stattdessen die Kachel und den Vorspann.
BIBER_AUSGEROTTET = 1869
BIBER_AUSSETZUNG = "1976–1982"

PERIODEN = ["2001–2006", "2007–2012", "2013–2018", "2019–2024"]

# Aus der Meldung 2019–2024, Spalte „Status Nature of change": In der
# alpinen Region tragen beide Arten `genuine` bei Wechsel U1 → FV. Das
# ist der einzige Satz dieses Abschnitts, der die Erholung nicht nur
# zeigt, sondern belegt — und er gehört deshalb in die Hinweiszeile.
ECHTE_ERHOLUNG = {
    "region": "Alpenraum",
    "arten": ["Biber", "Fischotter"],
    "von": "U1", "auf": "FV",
    "vermerk": "genuine",
}

# Untergrenze und Obergrenze der gemeldeten Individuenzahl, je Periode.
# None heißt: für diese Periode wurde keine Individuenzahl gemeldet.
ARTEN = [
    {
        "name": "Biber",
        "lateinisch": "Castor fiber",
        "spannen": [(2575, 2890), (4650, 4950), (7100, 7800), (13833, 16654)],
        "tiefpunkt": (
            f"{BIBER_AUSGEROTTET} war der Biber in Österreich ausgerottet. "
            f"Zwischen {BIBER_AUSSETZUNG} wurden an drei Stellen wieder "
            f"Tiere ausgesetzt."
        ),
    },
    {
        "name": "Fischotter",
        "lateinisch": "Lutra lutra",
        "spannen": [(430, 800), (950, 1350), (None, None), (3100, 4700)],
        "tiefpunkt": (
            "Der Fischotter war nie ganz verschwunden, erreichte aber in den "
            "1970er- und 1980er-Jahren seinen Tiefstand und lebte im Jahr "
            "2000 noch auf rund einem Fünftel der Bundesfläche."
        ),
    },
]


def baue_rueckkehrer() -> dict:
    log("\n[10/11] Biber und Fischotter — Rückkehrer (gepflegt)")

    arten = []
    for art in ARTEN:
        werte = []
        for periode, (unten, oben) in zip(PERIODEN, art["spannen"]):
            if unten is None or oben is None:
                werte.append({"periode": periode, "unten": None,
                              "oben": None, "spanne": None})
                continue
            if oben < unten:
                warnen(
                    f"{art['name']} {periode}: Obergrenze {oben} liegt unter "
                    f"der Untergrenze {unten} — Zahlen vertauscht?"
                )
            werte.append({"periode": periode, "unten": unten, "oben": oben,
                          "spanne": oben - unten})

        besetzt = [w for w in werte if w["unten"] is not None]
        erste, letzte = besetzt[0], besetzt[-1]
        # Faktor auf der Untergrenze gerechnet, nicht auf der Mitte: das
        # ist die vorsichtigste Lesart des Zuwachses.
        faktor = round(letzte["unten"] / erste["unten"], 1)

        arten.append({
            **{k: v for k, v in art.items() if k != "spannen"},
            "werte": werte,
            "erste_periode": erste["periode"],
            "letzte_periode": letzte["periode"],
            "erste_unten": erste["unten"], "erste_oben": erste["oben"],
            "letzte_unten": letzte["unten"], "letzte_oben": letzte["oben"],
            "faktor": faktor,
            "luecken": [w["periode"] for w in werte if w["unten"] is None],
        })
        log(f"    {art['name']}: {erste['unten']}–{erste['oben']} "
            f"({erste['periode']}) auf {letzte['unten']}–{letzte['oben']} "
            f"({letzte['periode']}) · Faktor {faktor}")

    pflegepruefung("rueckkehrer", STAND_JAHR, "Artikel-17-Meldung Säugetiere")

    quelle_vermerken(
        name="Eionet Artikel-17-Berichtswerkzeug — Österreich, Säugetiere",
        url=("https://nature-art17.eionet.europa.eu/article17/species/report/"
             "?period=6&group=Mammals&country=AT&region="),
        lizenz="Europäische Umweltagentur, Weiterverwendung erlaubt",
        stand=PERIODE,
        art="gepflegt",
    )

    return {
        "arten": arten,
        "perioden": PERIODEN,
        "periode": PERIODE,
        "stand": STAND_JAHR,
        "biber_ausgerottet": BIBER_AUSGEROTTET,
        "biber_aussetzung": BIBER_AUSSETZUNG,
        "echte_erholung": ECHTE_ERHOLUNG,
        "pflege": {
            "art": "gepflegt",
            "quelle": (
                "Eionet, Article 17 Web Tool: Meldungen Österreichs nach "
                "Artikel 17 FFH-Richtlinie, Säugetiere, Berichtsperioden "
                "2001–2006 bis 2019–2024. Gegenproben: BOKU Wildbiologie "
                "(Biber), Land Niederösterreich, Fischotter-Kartierung "
                "2022/23."
            ),
            "bericht_jahr": STAND_JAHR,
            "abgerufen": "2026-08-26",
        },
        "hinweis": (
            # 230 Zeichen, Fenster 150–234. Die Lücke beim Fischotter steht
            # bewusst nicht hier, sondern an der leeren Periode im Diagramm.
            "Die Zahlen sind Hochrechnungen aus Revierkartierungen, keine "
            "Zählungen — deshalb die Spanne. Im Alpenraum meldet Österreich "
            "für beide Arten eine echte Erholung, nicht bloß besseres Wissen. "
            "Seit 2019 ist begrenzte Entnahme erlaubt."
        ),
    }

File: docs/js/charts/rueckkehrer.js
/* ===========================================================================
   Biodiversitäts-Dashboard Österreich — Themenstrang: rueckkehrer
   ---------------------------------------------------------------------------
   Wird nach js/kern.js geladen; die Helfer kommen aus window.BIO.
   =========================================================================== */
(function (BIO) {
"use strict";
const { stil, zahl, pz, basis, achse, tabelle, setzeText, setzeHtml,
        diagramme, schrift, istSchmal, istEng, balkenGitter, kategorieLabel,
        balkenHoehe, legende, hoverDunkler } = BIO;

/* --- 10 — Biber und Fischotter, der Erholungspol ----------------------
   Liegende Balken, vier Berichtsperioden, zwei Arten.

   WARUM SPANNEN UND KEIN MITTELWERT: Beide Arten werden über Reviere und
   Nachweise erhoben und auf Individuen hochgerechnet. Eine einzelne Zahl
   wäre eine Genauigkeit, die die Erhebung nicht hergibt. Der Balken läuft
   deshalb von der gemeldeten Unter- zur Obergrenze und beginnt nicht am
   Nullpunkt.

   WIE EIN SCHWEBENDER BALKEN IN ECHARTS ENTSTEHT: Es gibt keinen
   Balkentyp, der von a nach b läuft. Jede Art bekommt deshalb ZWEI
   Reihen im selben Stapel — einen unsichtbaren Sockel bis zur
   Untergrenze und darüber die sichtbare Spanne. Zwei verschiedene
   `stack`-Namen sorgen dafür, dass die Arten NEBENEINANDER stehen statt
   übereinander. Die Sockel tragen `silent` und stehen nicht in der
   Legende; wer sie dort einträgt, bietet dem Leser ein Element an, das
   nichts bedeutet.

   DAS JAHR 1869 STEHT NICHT AUF DER ACHSE. Der Biber war damals in
   Österreich ausgerottet, die Aussetzung lief 1976–1982. Auf einer
   Kategorieachse wäre der Abstand 1869 → 2001 genauso breit wie sechs
   Jahre — eine Zeitlüge. Die Null trägt deshalb der Nachsatz unter der
   Grafik.

   DIE LÜCKE BEIM FISCHOTTER IST KEINE LÜCKE IN DER NATUR. Österreich
   meldete für 2013–2018 keine Individuen, sondern besetzte Rasterzellen.
   Diese Zahlen sind mit den Individuenzahlen davor und danach nicht
   vergleichbar. Sie werden deshalb NICHT umgerechnet — der Balken fehlt,
   und der Tooltip sagt, warum. Wer die Rasterzellen einsetzt, erzeugt
   einen Einbruch, den es nie gab. */

function baueRueckkehrer(daten) {
  const S = schrift();
  if (!daten?.arten?.length) return;
  const abschnitt = document.getElementById("s-rueckkehrer");
  if (abschnitt) abschnitt.style.display = "";

  const feld = document.getElementById("c-rueckkehrer");
  if (!feld) return;
  const d = echarts.getInstanceByDom(feld) || echarts.init(feld, null, { renderer: "svg" });
  if (!diagramme.includes(d)) diagramme.push(d);

  const perioden = daten.perioden;
  const arten = daten.arten;
  const farben = [stil("--viz-series-1"), stil("--viz-series-2")];
  const letzte = perioden.length - 1;

  setzeText("u-rueckkehrer",
    `Gemeldete Bestandsspanne je Berichtsperiode · ` +
    `${perioden[0]} bis ${daten.periode}`);
  setzeText("h-rueckkehrer", daten.hinweis ?? "");

  /* KORREKTUR 26.08.2026, Befund des Users: Hier stand die Null von 1869
     als große Zahl. Unter der Überschrift „sind zurückgekommen" liest
     sich eine riesige 0 als Widerspruch — sie ist der Wert von 1869 und
     nicht der heutige, und der Abschnitt handelt von zwei Arten, nicht
     von einer. Das Argument dafür war eines über Genauigkeit (die Null
     ist exakt, der heutige Bestand eine Spanne) und keines darüber, ob
     man es beim Lesen versteht.

     Jetzt trägt der Faktor die Zahl: Er misst dasselbe wie die Grafik —
     Wachstum über die Berichtsperioden — und ist auf der UNTERGRENZE
     gerechnet, also die vorsichtigste Lesart. 1869 bleibt als Pointe im
     Satz, wo es keinen Wert behauptet, den die Grafik nicht zeigt. */
  const biberPlakat = arten.find((a) => a.name === "Biber") || arten[0];
  const otter = arten.find((a) => a.name === "Fischotter");
  setzeHtml("k-rueckkehrer",
    `<span class="viz-plakat-zahl">${pz(biberPlakat.faktor)}` +
    `<span class="viz-plakat-einheit">×</span></span>` +
    `<p class="viz-plakat-satz">mehr Biber als in der Berichtsperiode ` +
    `${biberPlakat.erste_periode}. ${daten.biber_ausgerottet} war die Art ` +
    `in Österreich ausgerottet; heute leben hier wieder ` +
    `${zahl(biberPlakat.letzte_unten)} bis ${zahl(biberPlakat.letzte_oben)} ` +
    `Biber` +
    (otter ? ` und ${zahl(otter.letzte_unten)} bis ` +
             `${zahl(otter.letzte_oben)} Fischotter` : "") + `.</p>`);

  balkenHoehe(d, feld, perioden.length * arten.length, 30);

  /* BALKENBREITE — KORREKTUR 26.08.2026, Befund des Users am Bildschirmfoto.
     Hier stand `balkenBreite(feld, "62%")` wie in den übrigen Modulen. Dort
     ist es richtig, weil dort EINE Balkengruppe je Kategorie steht: 62 % der
     Bandbreite, 38 % Luft. Dieses Modul hat als einziges ZWEI Gruppen je
     Kategorie (`s0` und `s1` sind getrennte Stapel, damit die Arten
     nebeneinander stehen). Zwei Gruppen zu je 62 % ergeben 124 % — die
     Gruppe ist breiter als ihr Band, ECharts zentriert sie trotzdem, und
     jeder Balken rutscht aus seiner Zeile.

     Gemessen am gerenderten SVG bei 1.058 px Feldbreite: Band 58 px, Balken
     36 px, Bandmitten bei y = 63/121/179/237. Der Fischotter-Balken der
     Periode 2001–2006 lag bei y = 64,8–100,8 und damit auf der Beschriftung
     „2007–2012". Die Grafik ordnete jede Zahl der falschen Periode zu.

     Die Obergrenze für zwei Gruppen liegt bei 50 % minus Zwischenraum.
     40 % ergibt 23,2 px je Balken, dazu der ECharts-Standardabstand von
     30 % der Balkenbreite (7 px) — zusammen 53,4 px in einem 58-px-Band.

     ENG (unter 768 px): dort ist die Breite ein fester Pixelwert, weil der
     Kategoriename ÜBER dem Balken steht. `kategorieLabel` setzt dessen
     Unterkante auf `BAR_ENG / 2 + 4` = 11 px über die Bandmitte. Der feste
     Wert BAR_ENG = 14 aus `balkenBreite()` gilt für EINE Gruppe; zwei
     Gruppen zu 14 px belegen ±16 px und schöben den oberen Balken 5 px in
     den Namen. Deshalb 8 px: gemessen bei 700 px Fensterbreite belegt die
     Gruppe y = 66,35–83,15 um die Bandmitte 74,75, die Namensunterkante
     liegt bei 63,75 — 2,6 px Luft. Wer diesen Wert erhöht, schiebt den
     Namen in den oberen Balken. */
  const spannenBreite = istEng(feld) ? 8 : "40%";

  /* Je Art ein unsichtbarer Sockel und die sichtbare Spanne, in einem
     eigenen Stapel. `wert()` liest aus der Periodenliste der Art. */
  const reihen = [];
  arten.forEach((art, i) => {
    const wert = (k) => art.werte.find((w) => w.periode === perioden[k]) || {};
    reihen.push({
      name: `${art.name} (Sockel)`, type: "bar", stack: `s${i}`,
      silent: true, legendHoverLink: false,
      barWidth: spannenBreite,
      itemStyle: { color: "transparent" },
      emphasis: { disabled: true },
      data: perioden.map((p, k) => wert(k).unten ?? null),
    });
    reihen.push({
      name: art.name, type: "bar", stack: `s${i}`,
      barWidth: spannenBreite,
      itemStyle: { color: farben[i], borderRadius: 4 },
      emphasis: hoverDunkler(farben[i]),
      data: perioden.map((p, k) => {
        const w = wert(k);
        return w.spanne === null || w.spanne === undefined ? null : w.spanne;
      }),
      /* Beschriftet wird nur die jüngste Periode. Alle vier zu
         beschriften ergibt acht Zahlenpaare auf engem Raum; die
         übrigen Werte stehen in Tooltip und Tabelle. */
      label: {
        show: true, position: "right", distance: 8,
        color: stil("--viz-text-2"), fontSize: S.label,
        formatter: (p) => {
          if (p.dataIndex !== letzte) return "";
          const w = wert(letzte);
          if (w.unten === null || w.unten === undefined) return "";
          return `${zahl(w.unten)}–${zahl(w.oben)}`;
        },
      },
    });
  });

  d.setOption({
    ...basis(),
    grid: { ...balkenGitter(feld, { left: 96, right: 112 }), top: 34 },
    legend: legende(feld, {
      top: 0, left: istSchmal(feld) ? 0 : 96,
      itemWidth: 11, itemHeight: 11, itemGap: 16,
      data: arten.map((a) => a.name),
      textStyle: { color: stil("--viz-text-2"), fontSize: S.serie },
    }),
    tooltip: {
      ...basis().tooltip, trigger: "axis",
      axisPointer: { type: "shadow", shadowStyle: { color: stil("--viz-grid"), opacity: 0.35 } },
      formatter: (p) => {
        const k = p[0].dataIndex;
        const zeilen = [`<strong>${perioden[k]}</strong>`];
        arten.forEach((art, i) => {
          const w = art.werte.find((x) => x.periode === perioden[k]) || {};
          if (w.unten === null || w.unten === undefined) {
            zeilen.push(
              `<span style="color:${farben[i]}">■</span> ${art.name} ` +
              `<span style="color:${stil("--viz-muted")}">keine Individuenzahl ` +
              `gemeldet — Österreich meldete für diese Periode Rasterzellen</span>`);
          } else {
            zeilen.push(
              `<span style="color:${farben[i]}">■</span> ${art.name} ` +
              `<strong>${zahl(w.unten)}–${zahl(w.oben)}</strong>`);
          }
        });
        return zeilen.join("<br>");
      },
    },
    xAxis: { ...achse(), type: "value", min: 0, axisLine: { show: false },
      axisLabel: { hideOverlap: true, color: stil("--viz-muted"),
                   fontSize: S.achse, formatter: (v) => zahl(v) } },
    yAxis: { ...achse(), type: "category", inverse: true,
      data: perioden, splitLine: { show: false },
      axisLabel: { color: stil("--viz-text-2"),
                   fontSize: istSchmal(feld) ? S.eng : S.serie, margin: 12,
                   /* Viertes Argument: dieses Modul setzt seine Balkenhoehe eng selbst
       (`spannenBreite` = 8 px), holt sie also nicht von `balkenBreite()`.
       Die 14 halten den gemessenen Abstand von 2,6 px zwischen Name und
       Balken — siehe den Block ueber `spannenBreite`. Ohne das Argument
       zoege die Staffelung nach Zeilenzahl den Namen 7 px hoeher. */
    ...kategorieLabel(feld, 96, perioden.length, 14) } },
    series: reihen,
  }, { replaceMerge: ["series", "xAxis", "yAxis", "legend"] });

  /* Der Nachsatz trägt die beiden Zahlen, die im Balken nicht stehen
     können: die Null von 1869 und den Befund, dass die Erholung im
     Alpenraum als echt gemeldet ist und nicht als besseres Wissen. */
  const biber = arten.find((a) => a.name === "Biber") || arten[0];
  const saetze = [
    `<strong>${daten.biber_ausgerottet}</strong> war der Biber in Österreich ` +
    `ausgerottet. Zwischen ${daten.biber_aussetzung} wurden an drei Stellen ` +
    `wieder Tiere ausgesetzt — heute sind es ${zahl(biber.letzte_unten)} bis ` +
    `${zahl(biber.letzte_oben)}.`,
  ];
  if (daten.echte_erholung) {
    const e = daten.echte_erholung;
    saetze.push(
      `Im ${e.region} meldet Österreich für beide Arten eine <strong>echte ` +
      `Erholung</strong>, nicht bloß besseres Wissen: Der Erhaltungszustand ` +
      `wechselte dort von „${e.von}“ auf „${e.auf}“.`);
  }
  setzeHtml("n-rueckkehrer", saetze.join(" "));

  /* Eine Zeile je Periode, eine Spalte je Art. */
  setzeHtml("t-rueckkehrer", tabelle(
    [{ titel: "Berichtsperiode", wert: (z) => z.periode },
     ...arten.map((art) => ({
       titel: art.name, num: true,
       wert: (z) => {
         const w = art.werte.find((x) => x.periode === z.periode) || {};
         return w.unten === null || w.unten === undefined
           ? "–" : `${zahl(w.unten)}–${zahl(w.oben)}`;
       },
     }))],
    perioden.map((p) => ({ periode: p }))
  ));
}

BIO.baueRueckkehrer = baueRueckkehrer;
})(window.BIO);
